libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -169,15 +169,13 @@ proc getOrCreatePeer*(
 proc handleData*(p: PubSub, topic: string, data: seq[byte]): Future[void] {.async.} =
   if topic notin p.topics: return # Not subscribed
 
-  for h in p.topics[topic].handler:
-    trace "triggering handler", topicID = topic
-    try:
-      await h(topic, data)
-    except CancelledError as exc:
-      raise exc
-    except CatchableError as exc:
-      # Handlers should never raise exceptions
-      warn "Error in topic handler", msg = exc.msg
+  # gather all futures without yielding to scheduler
+  var futs = p.topics[topic].handler.mapIt(it(topic, data))
+  futs = await allFinished(futs)
```

Comment:
`allFinished` here is redundant - this code would be better written as:
```
for f in futs:
  try:
    await f
  except CancelledError: 
   ...
```

using `fut.failed` like in the PR also changes the cancellation behaviour

Reply:
@arnetheduck well https://github.com/status-im/nim-libp2p/pull/471#discussion_r536246637 😄

Reply:
so: `p.topics[topic].handler.mapIt(it(topic, data))` _starts_ all work so that it can proceed concurrently - `allFinished` adds an extra future that completes when _all_ futures have finished but is ultimately indistinguishable from simply iterating over the started futures and awaiting them serially - there's no advantage to calling `allFinished` unless the work that is being performed has dependencies and you can't start awaiting any of the futures until all have finished

Reply:
Advantage in number of async steps to complete, so if you will call `await f` N times you will finish in N*2 async steps, if you call `allFinished(futs)` you can finish in 2 steps.

Reply:
So if you able to replace
```nim
for fut in futs:
  await fut
```
do it with `allFinished()`, `allFutures()`.

Reply:
Yep in fact I liked my original approach but we keep bouncing this topic a lot 😄 so I wanted to avoid any discussion and used allfinished altho I wasn't really happy about it.
Of course @cheatfate is right about the numbers of steps. but cancellation worries me with allfinished.

Reply:
Old approach do not cares about cancellation too, because if cancellation will happen only one (current) Future will be cancelled, all other Futures in array will not be cancelled...

Reply:
what is the best approach to deal with cancellations then in this case? any suggestions?

Reply:
If its important to notify all the Futures in `futs` - do it, if you do not care - ignore it.
```nim
  # gather all futures without yielding to scheduler
  trace "triggering handler", topicID = topic
  var futs = p.topics[topic].handler.mapIt(it(topic, data))
  try:
    futs = await allFinished(futs)
  except CancelledError:
    for fut in futs:
      if not(fut.finished):
        fut.cancel()
```

Reply:
@cheatfate Ah.. so the cancellation would arrive on `allFinished`? this makes it much easier... would that work even if a inner future is cancelled?